src/pages/[platform]/build-a-backend/data/connect-to-existing-data-sources/single-table-design/index.mdx

```diff
@@ -0,0 +1,486 @@
+
+Optionally, tables can have a **sort key** that enables searching and sorting within items that share the same partition key. While partition keys require exact value searches, sort keys allow for pattern searches.
+
+With only partition and sort keys, query possibilities are limited without data duplication. To address this, DynamoDB offers two types of indexes:
+
+- **Global secondary index (GSIs)**: An index with a partition key and a sort key that can be different from those on the base table. A global secondary index is considered "global" because queries on the index can span all of the data in the base table, across all partitions. A global secondary index has no size limitations and has its own provisioned throughput settings for read and write activity that are separate from those of the table.
+
+- **Local secondary index (LSI)**: An index that has the same partition key as the base table, but a different sort key. A local secondary index is "local" in the sense that every partition of a local secondary index is scoped to a base table partition that has the same partition key value. As a result, the total size of indexed items for any one partition key value can't exceed 10 GB. Also, a local secondary index shares provisioned throughput settings for read and write activity with the table it is indexing.
+
+There are key differences between LSI and GSI:
+
+| Feature | LSI	| GSI |
+|--------|------|-----|
+| Create/Delete	| At table creation/deletion | Anytime |
+| Size Limit | Up to 10 GB per partition | Unlimited |
+| Throughput | Shared with table | Separate throughput |
+| Key Type | Primary key only | Composite key (partition & sort key) |
+| Consistency Model	| Both eventual and strong consistency | Eventual consistency only |
+
+## Determining Data Access Requirements
+
+Relational database design focuses on normalization without considering data access patterns. In contrast, NoSQL schema design prioritizes the list of questions the application needs to answer. NoSQL databases offer less dynamic query flexibility compared to SQL, so defining data access patterns upfront is crucial.
+
+To determine data access patterns in new applications, user stories and use cases can help identify query types. For migrating existing applications, query logs can reveal typical queries used.
+
+In this example, the frontend application requires these queries:
+
+1. Get results for each race by racer ID.
+2. Get a list of races by class ID.
+3. Get the best performance by racer for a class ID.
+4. Get the list of top scores by race ID.
+5. Get the second-by-second performance by racer for all races.
+
+While implementing this design with multiple DynamoDB tables is possible, it's unnecessary and inefficient. A key goal in querying DynamoDB data is retrieving all required data in a single query request. This is a core concept in NoSQL databases, and single-table design simplifies data management and maximizes query throughput.
```

Comment:
The language makes me wonder why typical Amplify UX does it the  "unnecessary and inefficient" way. It might be worth exploring the tradeoffs of taking this approach. Or, if we really believe this is a superior approach, we should consider adding this feature more natively into our L3 constructs and client libraries.

Reply:
It's not widely used partly because:
1) Few people know it exists; Amazon, for whatever reason, hasn't made a huge effort to promote it despite using it themselves on amazon.com.
2) You must define your access patterns upfront; many new projects prefer flexibility over efficiency.
3) It can cause problems if you don't know what you're doing. Having everything on one table is undeniably better, but for novices, wrapping their heads around partitions and profile details, sharing the same virtual space as order history, for example, might be confusing.
4) Denormalisation: This pattern sometimes requires repeated data, which can be challenging considering most people are familiar with relational normalised data.

That being said, it would be nice to see some native code for it, most people who follow this pattern currently define a table externally and use [ElectroDB](https://electrodb.dev/en/core-concepts/introduction/) within a lambda.